### PR TITLE
Using issue_comment even to trigger builds

### LINKED ISSUE #5073: Implement RFC 168 - Process Github issue_comment events to support adhoc task creation
Details in the RFC [here](https://github.com/taskcluster/taskcluster-rfcs/blob/main/rfcs/0168-Trigger-Tests-Based-on-PR-Comments.md).

### PR DESCRIPTION
**Pull request by lotas**
[RFC 168](https://github.com/taskcluster/taskcluster-rfcs/blob/main/rfcs/0168-Trigger-Tests-Based-on-PR-Comments.md) implementation

Github api now accepts `policy.issue_comment` webhook events (which needs new permission to be granted for an app and projects)

This event is being dispatched in the same way as the `pull_request` event. 

Delete event is being ignored, while created/edited are being equally accepted. (Note: If auto-cancellation of previous builds wasn't disabled, it would cancel previous runs on each comment edit, and create new one)

If comment doesn't belong to the pull request, it is ignored.

If comment body doesn't contain any line with `^/taskcluster\s+(.+)$` it is ignored.

Everything to the right of the `/taskcluster` is a "command" that will be accessible in `.taskcluster.yml` variables.

If person leaving comment is not a contributor, it is ignored and comment is added to the PR.

If `.taskcluter.yml` of the default branch doesn't allow comments, it will be ignored and comment will be added to the PR.

Before tasks would be rendered, head commit of the branch will be fetched, so that all variables could be passed to jsone context.

`.taskcluster.yml` will be rendered with following additional variables:

* `tasks_for` - `github-issue-comment`
* `event.taskcluster_comment` - trimmed string of the `/taskclsuter (.+)` line (without `/taskcluster`)
* `event.pull_request` - data fetched for the given PR during event processing
* `event.comment` - original event data see [webhook#issue_comment](https://docs.github.com/en/webhooks/webhook-events-and-payloads#issue_comment)

Fixes #5073

### CODE REVIEW

**Comment by lotas** [comment]
@ahal @bhearsum do you guys want to test it on dev before it gets merged? 
https://github.com/taskcluster/tc-dev-integration-test/pull/17

File: changelog/issue-5073.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+audience: developers
+level: minor
+reference: issue 5073
+---
+
+Github service supports `issue_comment` events to trigger jobs through `/tasckluster param` comments in open Pull Requests.
+`.taskcluster.yml` in default branch should allow this with `allowComments: collaborators` value.
+Tasks would be rendered with `tasks_for = "github-issue-comment"` and `event.taskcluster_comment = param`
```
Comment (petemoore) [inline changelog/issue-5073.md:8]: Should we add a link to RFC 168?

**Comment by bhearsum** [comment]
> @ahal @bhearsum do you guys want to test it on dev before it gets merged? [taskcluster/tc-dev-integration-test#17](https://github.com/taskcluster/tc-dev-integration-test/pull/17)

I would love to! I will take a look today or tomorrow.

File: services/github/src/api.js [hunk 2]
```diff
@@ -319,6 +319,36 @@ builder.declare({
         msg.branch = body.ref.split('/').slice(2).join('/');
         break;
 
+      case EVENT_TYPES.ISSUE_COMMENT:
+        // Comments on PRs can trigger tasks, too
+        // For this to work, there should be a `/taskcluster cmd` in the comment
+        // Plus repository should have a `.taskcluster.yml` in default branch with
+        // "allowComments: collaborators" in it
+        // Message is being processed in the same way as PULL_REQUEST
+        // and missing data would be fetched from the PR
```
Comment (petemoore) [inline services/github/src/api.js:328]: nice explanation ❤️

File: services/github/src/handlers/job.js [hunk 3]
```diff
@@ -162,6 +163,62 @@ export async function jobHandler(message) {
...
+    let defaultBranchYml = await this.getYml({ instGithub, owner: organization, repo: repository, ref: defaultBranch });
+    if (!defaultBranchYml) {
+      debug(`${organization}/${repository} has no '.taskcluster.yml' at ${defaultBranch}. Skipping.`);
+      return;
+    }
+    const allowCommentsPolicy = this.getRepoAllowCommentsPolicy(defaultBranchYml);
+
+    const validCommentPolicies = Object.values(ALLOW_COMMENT_POLICIES);
+    if (!validCommentPolicies.includes(allowCommentsPolicy)) {
+      debug(`allowComments: "${allowCommentsPolicy}" policy does not allow comments. Allowed: ${allowCommentsPolicy}. Skipping.`);
```
Comment (petemoore) [inline services/github/src/handlers/job.js:179]: this looks suspicious

File: services/github/src/handlers/job.js [hunk 4]
```diff
@@ -162,6 +163,62 @@ export async function jobHandler(message) {
...
+      await this.createComment({
+        instGithub, organization, repository, pullNumber, sha, debug,
+        body: {
+          summary: 'No Taskcluster jobs started for comment',
+          details: [
+            'The `allowComments` configuration for this repository ',
+            '(in `.taskcluster.yml` on the default branch) does not allow ',
+            'starting tasks from comments.',
```
Comment (petemoore) [inline services/github/src/handlers/job.js:187]: `policy.allowComments`

File: services/github/src/handlers/job.js [hunk 5]
```diff
@@ -162,6 +163,62 @@ export async function jobHandler(message) {
...
+          ].join(''),
+        },
+      });
+      await this.addCommentReaction({
+        instGithub, organization, repository,
+        commentId: message.payload.body.comment.id,
+        reaction: 'confused',
```
Comment (petemoore) [inline services/github/src/handlers/job.js:194]: nice

File: services/github/schemas/v1/taskcluster-github-config.v1.yml [hunk 6]
```diff
@@ -36,6 +36,14 @@ properties:
       Automatically cancel running task group on a non-default branch.
       Useful for PRs that are updated frequently. Only the last build will be running.
     default: true
+  allowComments:
+    type: string
```
Comment (petemoore) [inline services/github/schemas/v1/taskcluster-github-config.v1.yml:40]: Let's move this into policy section as a new property.

File: services/github/src/api.js [hunk 7]
```diff
@@ -319,6 +319,36 @@ builder.declare({
         msg.branch = body.ref.split('/').slice(2).join('/');
         break;
 
+      case EVENT_TYPES.ISSUE_COMMENT:
+        // Comments on PRs can trigger tasks, too
+        // For this to work, there should be a `/taskcluster cmd` in the comment
+        // Plus repository should have a `.taskcluster.yml` in default branch with
+        // "allowComments: collaborators" in it
```
Comment (petemoore) [inline services/github/src/api.js:326]: This might need tweaking if we move the config section to `policy.allowComments`

**Review by petemoore — APPROVED** [review]

**Comment by bhearsum** [comment]
> > @ahal @bhearsum do you guys want to test it on dev before it gets merged? [taskcluster/tc-dev-integration-test#17](https://github.com/taskcluster/tc-dev-integration-test/pull/17)
> 
> I would love to! I will take a look today or tomorrow.

Is it possible for me to hook up one of my own repos to this instance? I tried visiting https://github.com/apps/taskcluster-dev-app and even hacking the URL a bit, but I can't get it installed.

File: services/github/schemas/v1/taskcluster-github-config.v1.yml (hunk 6 above)
Comment (lotas) [reply services/github/schemas/v1/taskcluster-github-config.v1.yml:40]: yeah, I agree. I missed that we already have `policy.pullRequests` with very similar purposes. Moving it there

File: services/github/src/handlers/job.js (hunk 4 above)
Comment (lotas) [reply services/github/src/handlers/job.js:187]: yes! thanks

**Comment by lotas** [comment]
> > > @ahal @bhearsum do you guys want to test it on dev before it gets merged? [taskcluster/tc-dev-integration-test#17](https://github.com/taskcluster/tc-dev-integration-test/pull/17)
> > 
> > 
> > I would love to! I will take a look today or tomorrow.
> 
> Is it possible for me to hook up one of my own repos to this instance? I tried visiting https://github.com/apps/taskcluster-dev-app and even hacking the URL a bit, but I can't get it installed.

this is wild.. I don't see "Install App" section in the settings :) something is wrong

**Review by petemoore — APPROVED** [review]

**Pull request merged**